checkbox-ng/plainbox/impl/secure/qualifiers.py

```diff
@@ -492,9 +495,25 @@ def select_jobs(job_list, qualifier_list):
     # As a separate feature, we might return a list of qualifiers that never
     # matched anything. That may be helpful for debugging.
     included_list = []
-    id_to_index_map = {job.id: index for index, job in enumerate(job_list)}
     included_set = set()
     excluded_set = set()
+
+    def _handle_vote(qualifier, job):
+        """
+        Update list and sets of included/excluded jobs based on their related
+        qualifiers.
+        """
+        vote = qualifier.get_vote(job)
+        if vote == IJobQualifier.VOTE_INCLUDE:
+            if job in included_set:
+                return
+            included_set.add(job)
+            included_list.append(job)
```

Comment:
I had the same idea, but the list is here for a reason: to keep the jobs in the order of discovery. If you use a set, they end up in a weird order (especially in your proposal where you return them "sorted").

Reply:
I just tried your code, and it fails with

```python
  File "/home/pieq/dev/work/checkbox/checkbox-ng/plainbox/impl/secure/qualifiers.py", line 576, in select_jobs
    return sorted(included - excluded)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: '<' not supported between instances of 'JobDefinition' and 'JobDefinition'
```

If I replace

```python
    return sorted(included - excluded)
```

with

```python
    return included - excluded
```

it works, but the outcome of `checkbox-cli list-bootstrapped com.canonical.certification::client-cert-desktop-22-04` is very different from the current version of Checkbox. With my version, only the bootstrapped job order is different (basically `com.canonical.certification::snap` might be called before `com.canonical.certification::dmi` in one case, and after in the other, but this is because Checkbox spits the bootstrapped jobs randomly). With your proposal, the outcome is super weird, with chunks like this (random jobs stuck in the middle of suspend/resume cycles):

```
com.canonical.certification::stress-tests/suspend_cycles_11_reboot2
com.canonical.certification::stress-tests/suspend_cycles_12_reboot2
com.canonical.certification::stress-tests/suspend_cycles_13_reboot2
com.canonical.certification::stress-tests/suspend_cycles_14_reboot2
com.canonical.certification::stress-tests/suspend_cycles_15_reboot2
com.canonical.certification::ethernet/detect
com.canonical.certification::touchscreen/multitouch-rotate
com.canonical.certification::after-suspend-touchscreen/multitouch-rotate
com.canonical.certification::snap_revision_info
com.canonical.certification::snapd/snap-refresh-snapd-snapd-to-stable-rev
com.canonical.certification::snapd/reboot-after-snap-refresh-snapd-snapd-to-stable-rev
com.canonical.certification::snapd/snap-verify-after-refresh-snapd-snapd-to-stable-rev
com.canonical.certification::snapd/snap-revert-snapd-snapd-from-stable-rev
com.canonical.certification::snapd/reboot-after-snap-revert-snapd-snapd-from-stable-rev
com.canonical.certification::snapd/snap-verify-after-revert-snapd-snapd-from-stable-rev
com.canonical.certification::thunderbolt3/insert
com.canonical.certification::thunderbolt3/remove
com.canonical.certification::after-suspend-thunderbolt3/remove
com.canonical.certification::init-boot-loop-data
com.canonical.certification::cold-boot-loop-reboot1
com.canonical.certification::cold-boot-loop-test1
com.canonical.certification::cold-boot-loop-reboot2
com.canonical.certification::cold-boot-loop-test2
com.canonical.certification::graphics/1_glxgears_fullscreen_Alder_Lake-UP3_GT2__Iris_Xe_Graphics_
com.canonical.certification::after-suspend-graphics/1_glxgears_fullscreen_Alder_Lake-UP3_GT2__Iris_Xe_Graphics_
com.canonical.certification::stress-tests/suspend_cycles_16_reboot2
com.canonical.certification::stress-tests/suspend_cycles_17_reboot2
com.canonical.certification::stress-tests/suspend_cycles_18_reboot2
com.canonical.certification::stress-tests/suspend_cycles_19_reboot2
com.canonical.certification::stress-tests/suspend_cycles_20_reboot2
com.canonical.certification::ethernet/hotplug-enp0s31f6
com.canonical.certification::cpu/scaling_test
com.canonical.certification::cpu/scaling_test-log-attach
com.canonical.certification::monitor/1_type-c_hdmi_Alder_Lake-UP3_GT2__Iris_Xe_Graphics_
com.canonical.certification::audio/external-linein
com.canonical.certification::stress-tests/suspend_cycles_21_reboot2
com.canonical.certification::stress-tests/suspend_cycles_22_reboot2
```

Reply:
FYI, the full diff between version in `main` and your proposal:

https://pastebin.ubuntu.com/p/qkFKh5ZCRG/

Reply:
The ordering was something I was afraid we need to preserve. In this case I'd love if you could drop a comment there (where the two collections are defined), so the next person understands why it's there.